### ISSUE TITLE
make error

### ISSUE THREAD

**Issue by yangchengwork** [cryptodev-linux/cryptodev-linux#46]
DKMS make.log for cryptodev-1.10 for kernel 4.4.176-1-ARCH (armv5tel)
Fri Aug 16 07:40:52 UTC 2019
/var/lib/dkms/cryptodev/1.10/build/zc.c: In function '__get_userbuf':
/var/lib/dkms/cryptodev/1.10/build/zc.c:64:44: warning: passing argument 7 of 'get_user_pages' from incompatible pointer type [-Wincompatible-pointer-types]
    (unsigned long)addr, pgcount, write, 0, pg, NULL);
                                            ^~
In file included from /var/lib/dkms/cryptodev/1.10/build/zc.c:28:
include/linux/mm.h:1225:31: note: expected 'struct vm_area_struct **' but argument is of type 'struct page **'
       struct vm_area_struct **vmas);
       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/var/lib/dkms/cryptodev/1.10/build/zc.c:63:8: error: too many arguments to function 'get_user_pages'
  ret = get_user_pages(task, mm,
        ^~~~~~~~~~~~~~
In file included from /var/lib/dkms/cryptodev/1.10/build/zc.c:28:
include/linux/mm.h:1222:6: note: declared here
 long get_user_pages(struct task_struct *tsk, struct mm_struct *mm,
      ^~~~~~~~~~~~~~
make[2]: *** [scripts/Makefile.build:278: /var/lib/dkms/cryptodev/1.10/build/zc.o] Error 1
make[1]: *** [Makefile:1436: _module_/var/lib/dkms/cryptodev/1.10/build] Error 2
make: *** [Makefile:27: build] Error 2

**Comment by qosmio** [comment]
Having the same issue. Not sure if you were able to make an progress since Aug?

**Comment by horiag** [comment]
Looks like this is due to commit 8e50b8b07f46 ("mm: replace get_user_pages() write/force parameters with gup_flags") from linux-stable/linux-4.4.y, included in release v4.4.168.

Linux kernel version checking in __get_userbuf() has to be updated.

**Comment by kingforaday** [comment]
Kernel 4.4.y doesn't implement `get_user_pages_remote()`.  As horiag mentioned, in 4.4.168 the force flag from get_user_pages was removed:
`@@ -1195,19 +1199,17 @@ long __get_user_pages(struct task_struct *tsk, struct mm_struct *mm,
 		      struct vm_area_struct **vmas, int *nonblocking);
 long get_user_pages(struct task_struct *tsk, struct mm_struct *mm,
 		    unsigned long start, unsigned long nr_pages,
-		    int write, int force, struct page **pages,
+		    unsigned int gup_flags, struct page **pages,
 		    struct vm_area_struct **vmas);`

Making the change to `zc.c` worked.
`diff --git a/zc.c b/zc.c
index 2c286bb..46c0ca1 100644
--- a/zc.c
+++ b/zc.c
@@ -65,7 +65,7 @@ int __get_userbuf(uint8_t __user *addr, uint32_t len, int write,
 #endif
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 6, 0))
        ret = get_user_pages(task, mm,
-                       (unsigned long)addr, pgcount, write, 0, pg, NULL);
+                       (unsigned long)addr, pgcount, write, pg, NULL);
 #elif (LINUX_VERSION_CODE < KERNEL_VERSION(4, 9, 0))
        ret = get_user_pages_remote(task, mm,
                        (unsigned long)addr, pgcount, write, 0, pg, NULL);`
